### PR TITLE
[SPARK-23635][YARN] AM env variable should not overwrite same name env variable set through spark.executorEnv.

### PR DESCRIPTION
**Pull request by jerryshao**
## What changes were proposed in this pull request?

In the current Spark on YARN code, AM always will copy and overwrite its env variables to executors, so we cannot set different values for executors.

To reproduce issue, user could start spark-shell like:

```
./bin/spark-shell --master yarn-client --conf spark.executorEnv.SPARK_ABC=executor_val --conf  spark.yarn.appMasterEnv.SPARK_ABC=am_val
```

Then check executor env variables by

```
sc.parallelize(1 to 1).flatMap \{ i => sys.env.toSeq }.collect.foreach(println)
```

We will always get `am_val` instead of `executor_val`. So we should not let AM to overwrite specifically set executor env variables.

## How was this patch tested?

Added UT and tested in local cluster.

### CODE REVIEW

**Comment by jerryshao** [comment]
@mridulm , would you please take a review. Thanks!

**Comment by mridulm** [comment]
Thanks for fixing this @jerryshao !
LGTM
Since there is a behavior change, +CC @tgravescs and @vanzin

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ExecutorRunnable.scala [hunk 1]
```diff
@@ -247,6 +241,18 @@ private[yarn] class ExecutorRunnable(
 
     System.getenv().asScala.filterKeys(_.startsWith("SPARK"))
       .foreach { case (k, v) => env(k) = v }
+
+    sparkConf.getExecutorEnv.foreach { case (key, value) =>
+      if (key == Environment.CLASSPATH.name()) {
```
Comment (tgravescs) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ExecutorRunnable.scala]: there are other variables that requires the append, like LD_LIBRARY_PATH.   It obviously can do bad things if its specified twice though but generally you shouldn't be doing that.  

Maybe I missed it in the description exactly how is AM env taking presendence?  Can you please describe the flow as to why moving this or changing it fixes the bug (how is executor getting am env now)?

Comment (mridulm) [reply resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ExecutorRunnable.scala]: @tgravescs In existing code, in `prepareEnvironment`, "env" is populated only with `Environment.CLASSPATH`. Hence LD_LIBRARY_PATH does not apply to this specific change.

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ExecutorRunnable.scala [hunk 2]
```diff
@@ -247,6 +241,18 @@ private[yarn] class ExecutorRunnable(
...
+        // If the key of env variable is CLASSPATH, we assume it is a path and append it.
+        // This is kept for backward compatibility and consistency with hadoop
+        YarnSparkHadoopUtil.addPathToEnvironment(env, key, value)
+      } else {
+        // For other env variables, simply overwrite the value.
+        env(key) = value
+      }
+    }
```
Comment (mridulm) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ExecutorRunnable.scala]: @jerryshao I think there is a potential issue with this change - it allows for users to (incorrectly) specify SPARK_LOG_URL_STDERR, SPARK_LOG_URL_STDOUT : which should be generated by driver. The section "// Add log urls" above this code snippet.

Note, this is an existing bug in the code regarding the same - if the same variables had been present in driver env, they would have overridden the generated value's.
Would be good to fix this issue as well as part of this change.

Solution would be to move the block for '// Add log urls' below this current block

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ExecutorRunnable.scala (hunk 1 above)
Comment (tgravescs) [reply resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ExecutorRunnable.scala]: Ah, I see sorry missed that. So I guess here we are just stomping on whatever is in the system env path now, vs before we were stomping on the executorEnv specified with the system env.

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ExecutorRunnable.scala (hunk 2 above)
Comment (jerryshao) [reply resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ExecutorRunnable.scala]: Sure, I will fix it, thanks for the reminder.

**Comment by mridulm** [comment]
Thanks for the changes @jerryshao, they look good to me.
Do you have any other thoughts/comments @tgravescs ? Since this is a behavior change, I want to make sure we are not missing anything !

**Comment by tgravescs** [comment]
looks good to me

**Comment by jerryshao** [comment]
Thanks for the review, let me merge to master.

**Pull request closed without merging**